cli/launchpad/folder.go

```diff
@@ -26,13 +26,27 @@ type folderSpecYAML struct { // Inner mappings
 type folders []*folderYAML
 
 // add appends a folder into the folder list if it does not exist already.
-func (fs *folders) add(newF *folderYAML) {
+//
+// add returns existing folder's reference if attempted to add folder of the same ID.
+func (fs *folders) add(newF *folderYAML) *folderYAML {
 	for _, f := range *fs {
-		if f.Spec.Id == newF.Spec.Id { // silently ignore already existing folder
-			return
+		if f.Spec.Id == newF.Spec.Id {
+			return f
 		}
 	}
 	*fs = append(*fs, newF)
+	return nil
+}
+
+// merge adds given list of folders into the current folders list recursively.
+//
+// merge will be called on subfolder when it exist in both current and given list of folders.
+func (fs *folders) merge(nfs *folders) {
+	for _, f := range *nfs {
+		if existF := fs.add(f); existF != nil {
```

Comment:
Shouldn't we *always* add? I don't necessarily see why `merge` should only happen on existing folders.

Reply:
Ah, this merge happens when we are joining two lists together. The only possible recursive call is when a folder exists in both list. ex:
`fs = []folder{f1, f2, f3}; nfs = []folder{f3, f4, f5}`
Expected output of `fs.merge(nfs)` would be `fs=[]folder{f1, f2, f3, f4, f5}`. 
Calling merge on `f3` is natural because `f3` in `fs` might have subfolder `f31, f32`, and `f3` in `nfs` might have `f32, f33`, which we expect eventually `f3` in `fs` to be `f31, f32, f33`. 
However, calling f1 to merge is unnecessary because `f1` doesn't exist in `nfs` and it's a unique copy. Similar on `f5` that could just be added onto `fs` without a merge operation.